### PR TITLE
Fix Initialization of FilteredIterator from base class

### PR DESCRIPTION
**Pull request by masterleinad**
Before this PR `mpi/step-39.mpirun=1.debug` fails due to 
>  FilteredIterator<BaseIterator>::operator = (const FilteredIterator &fi)

giving
> include/deal.II/grid/tria_iterator.h:872:14: error: no matching function for call to ‘dealii::DoFCellAccessor<dealii::DoFHandler<2>, true>::DoFCellAccessor(const dealii::FilteredIterator<dealii::TriaIterator<dealii::DoFCellAccessor<dealii::DoFHandler<2>, true> > >&)’
accessor (a)

An explicit upcast fixes this.

### CODE REVIEW

File: include/deal.II/grid/filtered_iterator.h [hunk 1]
```diff
@@ -960,10 +960,8 @@ FilteredIterator<BaseIterator> &
 FilteredIterator<BaseIterator>::
 operator = (const FilteredIterator &fi)
 {
-  Assert ((fi.state() != IteratorState::valid) || (*predicate)(fi),
-          ExcInvalidElement(fi));
```
Comment (masterleinad) [inline include/deal.II/grid/filtered_iterator.h]: The same checks are performed in `operator=(const BaseIterator &)` which is now called.

File: include/deal.II/grid/filtered_iterator.h [hunk 2]
```diff
@@ -960,10 +960,8 @@ FilteredIterator<BaseIterator> &
...
-  BaseIterator::operator = (fi);
-  return *this;
+  const BaseIterator &bi = fi;
+  return operator = (bi);
```
Comment (bangerth) [inline include/deal.II/grid/filtered_iterator.h]: What kind of error did you get from the existing code?

Comment (masterleinad) [inline include/deal.II/grid/filtered_iterator.h]: The last interesting lines are
> 6400: include/deal.II/grid/tria_iterator.h: In instantiation of ‘dealii::TriaRawIterator<Accessor>::TriaRawIterator(const OtherAccessor&) [with OtherAccessor = dealii::FilteredIterator<dealii::TriaIterator<dealii::DoFCellAccessor<dealii::DoFHandler<2>, true> > >; Accessor = dealii::DoFCellAccessor<dealii::DoFHandler<2>, true>]’:

> 6400: include/deal.II/grid/tria_iterator.h:1110:31:   required from ‘dealii::TriaIterator<Accessor>::TriaIterator(const OtherAccessor&) [with OtherAccessor = dealii::FilteredIterator<dealii::TriaIterator<dealii::DoFCellAccessor<dealii::DoFHandler<2>, true> > >; Accessor = dealii::DoFCellAccessor<dealii::DoFHandler<2>, true>]’

> 6400: include/deal.II/grid/filtered_iterator.h:963:3:   required from ‘dealii::FilteredIterator<BaseIterator>& dealii::FilteredIterator<BaseIterator>::operator=(const dealii::FilteredIterator<BaseIterator>&) [with BaseIterator = dealii::TriaIterator<dealii::DoFCellAccessor<dealii::DoFHandler<2>, true> >]’

with the error in the PR description. Basically an implicit upcast is not performed in favor of trying to initialize `dealii::DoFCellAccessor<dealii::DoFHandler<2>, true>` with `dealii::FilteredIterator<dealii::TriaIterator<dealii::DoFCellAccessor<dealii::DoFHandler<2>, true> > >`

Comment (bangerth) [inline include/deal.II/grid/filtered_iterator.h]: Hm, ok.

If I saw the code you have, I'd probably be tempted to just change it back to its original form. Do you have a test that one could add to make sure we don't accidentally change things back?

Comment (masterleinad) [inline include/deal.II/grid/filtered_iterator.h]: `mpi/step-39.mpirun=1.debug` is already failing but I can of course create another test that is more specific to this issue.

Comment (bangerth) [inline include/deal.II/grid/filtered_iterator.h]: Ah, I missed that you already have a failing testcase (which, incidentally, doesn't seem to fail for me, so it must be a problem with the compiler).

If it's easy to extract a small self-contained testcase, that would be appreciated. If not, let's not worry about it.

**Comment by masterleinad** [comment]
In addition to the test, I also added some comments.

**Comment by bangerth** [comment]
Much appreciated. OK to merge once the tester is happy.

**Pull request merged**